### ISSUE TITLE
LimitExceeded but I didn't have any instance deployed.

### ISSUE THREAD

**Issue by peprmint1** [hitrov/oci-arm-host-capacity#104]
`{
"code": "LimitExceeded",
"message": "The following service limits were exceeded: standard-a1-memory-count, standard-a1-memory-regional-count, standard-a1-core-count, standard-a1-core-regional-count. Request a service limit increase from the service limits page in the console. "
}`

I received this response several times for the last 14 days. According to the docs provided by this repository that kind of response stated that I already have an instance but I didn't have any instance, undeleted boot volume, etc. and I also didn't see any new instance deployed after I received this kind of response when running this script.

Is this the script's vault or Oci's or me? Please help, I'm so confused.

Thanks.

**Comment by peprmint1** [comment]
Yes I did checked it, all the free tier resources is not being used at all 
![image](https://github.com/hitrov/oci-arm-host-capacity/assets/63215668/20ee59c7-a26a-4308-9a19-a3a5951af6cd)

here is my .env config
![image](https://github.com/hitrov/oci-arm-host-capacity/assets/63215668/27e8513a-e7e8-4465-9e55-fc1cfa2c54f2)

I don't think my account was limited, any idea why.

Thank you for your response.

**Comment by hitrov** [comment]
Weird. Try creating from the console, and if error will be different (“normal” out of host capacity), send here copied curl request from the browser (without auth headers)

**Comment by peprmint1** [comment]
`curl 'https://iaas.ap-singapore-1.oraclecloud.com/20160918/instances/' \
  -H 'Accept: */*' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'DNT: 1' \
  -H 'Origin: https://cloud.oracle.com' \
  -H 'Referer: https://cloud.oracle.com/' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: cross-site' \
  -H 'User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36' \
  -H 'accept-language: en' \
  -H 'authorization: <removed> \
  -H 'opc-request-id: <removed>' \
  -H 'sec-ch-ua: "Google Chrome";v="119", "Chromium";v="119", "Not?A_Brand";v="24"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Windows"' \
  -H 'x-content-sha256: <removed> \
  -H 'x-date: Thu, 07 Dec 2023 12:50:08 GMT' \
  --data-raw '{"metadata":{"ssh_authorized_keys":<removed>},"shape":"VM.Standard.A1.Flex","compartmentId":"ocid1.tenancy.oc1..aaaaaaaaypub72axxtqi323k2wuuihpuq6llaee6ne5x7jwg5qxgyhw3vp5q","displayName":"instance-20231207-1948","availabilityDomain":"fgcZ:AP-SINGAPORE-1-AD-1","sourceDetails":{"sourceType":"image","imageId":"ocid1.image.oc1.ap-singapore-1.aaaaaaaaodapmrmktgbfke35cbewiajm5nlzlfdplbirozptgaiu5reoavoq","bootVolumeSizeInGBs":100,"bootVolumeVpusPerGB":10},"isPvEncryptionInTransitEnabled":true,"createVnicDetails":{"assignPublicIp":true,"subnetId":"ocid1.subnet.oc1.ap-singapore-1.aaaaaaaaco6ll3qeqyegfqr4675m5uoa5nmcuegh7qmj3ctgzzo54vtanj2a","assignPrivateDnsRecord":true,"assignIpv6Ip":true,"ipv6AddressIpv6SubnetCidrPairDetails":[{"ipv6SubnetCidr":"2603:c024:450b:b400:0000:0000:0000:0000/64"}]},"agentConfig":{"pluginsConfig":[{"name":"Vulnerability Scanning","desiredState":"DISABLED"},{"name":"Compute RDMA GPU Monitoring","desiredState":"DISABLED"},{"name":"Compute Instance Monitoring","desiredState":"ENABLED"},{"name":"Compute HPC RDMA Auto-Configuration","desiredState":"DISABLED"},{"name":"Compute HPC RDMA Authentication","desiredState":"DISABLED"},{"name":"Block Volume Management","desiredState":"DISABLED"},{"name":"Bastion","desiredState":"DISABLED"}],"isMonitoringDisabled":false,"isManagementDisabled":false},"definedTags":{},"freeformTags":{},"instanceOptions":{"areLegacyImdsEndpointsDisabled":false},"availabilityConfig":{"recoveryAction":"RESTORE_INSTANCE"},"shapeConfig":{"ocpus":2,"memoryInGBs":12}}' \
  --compressed`

Sure here it is.

**Comment by C0nvert** [comment]
Can confirm I get the same error code.

**Comment by cuzz1369** [comment]
I get both, I get 2 or 3 "Out of Capacity" in a row, then 1 or 2 "Limit Exceeded" in a row. i run mine every minute.

**Comment by hitrov** [comment]
my guess it could be related to the number of API calls allowed for certain endpoint (create instance). everywhere I recommend to not run every minute, at least once per 3-5

**Comment by cuzz1369** [comment]
OK, I will space it out to every 3 minutes and see if that helps. Thanks.

**Comment by peprmint1** [comment]
Now my account is blocked 🙂. I run this script every 5mins.

**Comment by hitrov** [comment]
weird, regret to gear that. did you have any AMD instances or other resources which could possibly violate their policy?
the worst thing is that Oracle REST API doesn't provide any rate limit (-remaining) headers to not even get into this situation. the best thing that I could I already did many months ago by implementing [Add too many requests waiter](https://github.com/hitrov/oci-arm-host-capacity/commit/4a584c413ab1ea58345a7311d370bf449ea241a3) - basically do nothing for the next ten minutes after `Too Many Requests` HTTP error code encountered

**Comment by peprmint1** [comment]
Nope, just a VCN no instances, no volumes, no buckets, just a VCN. I've been using this script to get an instance since Feb 2022 and it worked several times and this script is very helpful until my account got blocked. I think OCI do something with their ends to detect requests that comes from this script and blocked them.

**Comment by peprmint1** [comment]
Is there any way to sign up again to OCI, I tried to sign up but always got "Error processing transaction" after verifying my payment method (got charged) and submitted them. I have contacted their support and they just reply "Sorry we're unable to process... This is all the information we can provide".

**Comment by hitrov** [comment]
That would be strange because official REST API is used, there are no any violations that I could think of

**Comment by hitrov** [comment]
In fact you pointed in a way that maybe I’m gonna check official SDK implementation, maybe some more headers can be added to “pretend” as there’s no PHP one, I had to write auth on my own https://github.com/hitrov/oci-api-php-request-sign

**Comment by peprmint1** [comment]
> That would be strange because official REST API is used, there are no any violations that I could think of

Indeed, don't get me wrong but even their payment verification system is faulty and before my account got blocked I have tried to upgrade to Pay As You Go but I can't get through the US$100 hold because they said my card was denied. I contacted my bank they said nothing was wrong the international transaction was enabled and the card was not at its limit. The bank was right I can use the same card on Stripe and PayPal but not on OCI, tried several different cards issued from completely different banks still face the same issue, the cards was all denied.

**Comment by cuzz1369** [comment]
> my guess it could be related to the number of API calls allowed for certain endpoint (create instance). everywhere I recommend to **not** run every minute, at least once per 3-5

It was on me, I figured out shortly after my last comment that I had 2 crons running. Stopped one and spaced out to every 3 minutes and it has been fine since.

**Comment by divinity76** [comment]
Fwiw i ran this script in a loop every 700/1000 seconds, and i finally got my free server! it took 1 day and 14 hours:
```
real    2303m58.653s
user    0m20.932s
sys     0m18.722s
```
and the runner script I used:
```php
<?php

declare(strict_types=1);
/**
 * better version of shell_exec(),
 * supporting both stdin and stdout and stderr and os-level return code
 *
 * @param string $cmd
 *            command to execute
 * @param string $stdin
 *            (optional) data to send to stdin, binary data is supported.
 * @param string $stdout
 *            (optional) stdout data generated by cmd
 * @param string $stderr
 *            (optional) stderr data generated by cmd
 * @param bool $print_std
 *            (optional, default false) if you want stdout+stderr to be printed while it's running,
 *            set this to true. (useful for long-running commands)
 * @return int
 */
function hhb_exec(string $cmd, string $stdin = "", string &$stdout = null, string &$stderr = null, bool $print_std = false): int
{
    $stdouth = tmpfile();
    $stderrh = tmpfile();
    $descriptorspec = array(
        0 => array(
            "pipe",
            "rb"
        ), // stdin
        1 => array(
            "file",
            stream_get_meta_data($stdouth)['uri'],
            'ab'
        ),
        2 => array(
            "file",
            stream_get_meta_data($stderrh)['uri'],
            'ab'
        )
    );
    $pipes = array();
    $proc = proc_open($cmd, $descriptorspec, $pipes);
    while (strlen($stdin) > 0) {
        $written_now = fwrite($pipes[0], $stdin);
        if ($written_now < 1 || $written_now === strlen($stdin)) {
            // ... can add more error checking here
            break;
        }
        $stdin = substr($stdin, $written_now);
    }
    fclose($pipes[0]);
    unset($stdin, $pipes[0]);
    if (!$print_std) {
        $proc_ret = proc_close($proc); // this line will stall until the process has exited.
        $stdout = stream_get_contents($stdouth);
        $stderr = stream_get_contents($stderrh);
    } else {
        $stdout = "";
        $stderr = "";
        stream_set_blocking($stdouth, false);
        stream_set_blocking($stderrh, false);
        $fetchstd = function () use (&$stdout, &$stderr, &$stdouth, &$stderrh): bool {
            $ret = false;
            $tmp =  stream_get_contents($stdouth); // fread($stdouth, 1); //
            if (is_string($tmp) && strlen($tmp) > 0) {
                $ret = true;
                $stdout .= $tmp;
                fwrite(STDOUT, $tmp);
            }
            $tmp = stream_get_contents($stderrh); // fread($stderrh, 1); //
            // var_dump($tmp);
            if (is_string($tmp) && strlen($tmp) > 0) {
                $ret = true;
                $stderr .= $tmp;
                fwrite(STDERR, $tmp);
            }
            return $ret;
        };
        while (($status = proc_get_status($proc))["running"]) {
            if (!$fetchstd()) {
                // 100 ms
                usleep(100 * 1000);
            }
        }
        $proc_ret = $status["exitcode"];
        proc_close($proc);
        $fetchstd();
    }
    fclose($stdouth);
    fclose($stderrh);
    return $proc_ret;
}

$cmd = "php index.php 2>&1";
for (;;) {
    echo ".";
    $ret = hhb_exec($cmd, "", $stdout, $stderr, true);
    $expected = array(
        'ret' => 0,
        'stdout' => '{
          "code": "InternalError",
          "message": "Out of host capacity."
      }
      {
          "code": "InternalError",
          "message": "Out of host capacity."
      }
      {
          "code": "InternalError",
          "message": "Out of host capacity."
      }
      ',
        'stderr' => '',
    );
    $expected['stdout'] = trim($expected['stdout']);
    $actual = array(
        'ret' => $ret,
        'stdout' => $stdout,
        'stderr' => $stderr,
    );
    $actual['stdout'] = trim($actual['stdout']);
    if (
        str_starts_with($actual['stdout'], 'PHP Fatal error:  Uncaught Hitrov\\Exception\\TooManyRequestsWaiterException: Will retry after')
        || str_contains($actual['stdout'], '"TooManyRequests"')
    ) {
        // Will retry after 83 seconds
        preg_match('/Will retry after (\d+) seconds/', $actual['stdout'], $matches);
        //var_dump($matches);
        $seconds = (int) ($matches[1] ?? 100);
        $seconds = 1000;
        for ($i = 1; $i < ($seconds + 1); ++$i) {
            echo "{$i}/{$seconds}\r";
            sleep(1);
        }
        echo "\n";
        continue;
    }
    if (
        preg_replace('/\s+/', '', $expected['stdout']) !== preg_replace('/\s+/', '', $actual['stdout'])
    ) {
        echo "Unexpected STDOUT output, dumping...\n";
        $actual = var_export($actual, true);
        $expected = var_export($expected, true);
        file_put_contents(__FILE__ . ".actual", $actual);
        file_put_contents(__FILE__ . ".expected", $expected);
        echo "Actual output:\n$actual\n";
        die();
    }
    if ($expected['stderr'] !== $actual['stderr']) {
        echo "Unexpected STDERR output, dumping...\n";
        $actual = var_export($actual, true);
        $expected = var_export($expected, true);
        file_put_contents(__FILE__ . ".actual", $actual);
        file_put_contents(__FILE__ . ".expected", $expected);
        echo "Actual output:\n$actual\n";
        die();
    }
    if ($expected['ret'] !== $actual['ret']) {
        echo "Unexpected return code, dumping...\n";
        $actual = var_export($actual, true);
        $expected = var_export($expected, true);
        file_put_contents(__FILE__ . ".actual", $actual);
        file_put_contents(__FILE__ . ".expected", $expected);
        echo "Actual output:\n$actual\n";
        die();
    }
    for($i=1,$imax=701;$i<$imax; ++$i){
        echo "{$i}/{$imax}\r";
        sleep(1);
    }
    continue;
    echo "Unexpected output, dumping...\n";
    $actual = var_export($actual, true);
    $expected = var_export($expected, true);
    file_put_contents(__FILE__ . ".actual", $actual);
    file_put_contents(__FILE__ . ".expected", $expected);
    echo "Actual output:\n$actual\n";
    die();
}
```

**Comment by peprmint1** [comment]
@divinity76 Nope, there are no boot volumes.

**Comment by vasmarfas** [comment]
I have script, that run every  minute on 2 machines (nowm after i read  about banned accounts, i set to every 8 min on first machine and every 12 min on second machine). And i getting the same error every 20-90 minutes.  20-90 minutes i got every minute standart `Out of host capacity.` and once in this time i got `The following service limits were exceeded: standard-a1-memory-count, standard-a1-memory-regional-count, standard-a1-core-count, standard-a1-core-regional-count. Request a service limit increase from the service limits page in the console. `. All my limits are ok:
<img width="735" alt="image" src="https://github.com/hitrov/oci-arm-host-capacity/assets/37144504/94529843-9c9d-494d-87b8-b9338c3e8d56">
<img width="772" alt="image" src="https://github.com/hitrov/oci-arm-host-capacity/assets/37144504/2c710df8-7608-4270-a569-12d6418cae8a">
<img width="759" alt="image" src="https://github.com/hitrov/oci-arm-host-capacity/assets/37144504/17a13e9e-0fcc-43fc-ab71-6dc545ca5987">

**Comment by divinity76** [comment]
@vasmarfas running with the same api key on 2 different computers? Sounds like a dangerous game IMO.
